### PR TITLE
Tests for puppet resource ec2_instance

### PR DESCRIPTION
**Pull request by zreichert**
All updates made.

### CODE REVIEW

File: spec/acceptance/instance_spec.rb [hunk 1]
```diff
@@ -1,11 +1,13 @@
 require 'spec_helper_acceptance'
 require 'securerandom'
 require 'ipaddr'
+require 'pry'
```
Comment (garethr) [inline spec/acceptance/instance_spec.rb]: probably best left out, only used for debugging and might not be in the Gemfile for running the acceptance tests

File: spec/acceptance/instance_spec.rb [hunk 2]
```diff
@@ -195,4 +197,153 @@ def expect_failed_apply(config)
...
+      end
+      # create new instance with puppet resource
+      TestExecutor.puppet_resource('ec2_instance', @config, '--modulepath ../')
+      #wait for instance to report as running
+      @instance = get_instance(@config[:name])
+      @aws.ec2_client.wait_until(:instance_running, instance_ids:[@instance.instance_id])
+      # set env variable and use puppet resource to inspect state of ec2 instance
+      ENV['AWS_REGION'] = @config[:region]
+      @result = TestExecutor.puppet_resource('ec2_instance', {:name => @config[:name]}, '--modulepath ../')
+      expect(@result.stderr).not_to match(/Error:/)
```
Comment (garethr) [inline spec/acceptance/instance_spec.rb]: Could this just check for anything in stderr? Or the exit code?

File: spec/acceptance/instance_spec.rb [hunk 3]
```diff
@@ -195,4 +196,153 @@ def expect_failed_apply(config)
...
+        :image_id => 'ami-41e85d5c',
+        :ensure => 'present',
+        :monitoring => false,
+        :availability_zone => @default_availability_zone,
+        :security_groups => 'default',
+      }
+
+      # The value for this ENV var must be an existing key in your Amazon account
+      # the key must be available in the sa-east-1 region
+      if ENV['AWS_KEY_PAIR']
```
Comment (garethr) [inline spec/acceptance/instance_spec.rb]: super picky, I'd probably collapse this down to one line

```ruby
@config[:key_name] = ENV['AWS_KEY_PAIR'] if ENV['AWS_KEY_PAIR']
```

**Comment by garethr** [comment]
This looks good to me. I'm happy to merge once @Iristyle or @cmurphy have given the go on #52

File: spec/acceptance/instance_spec.rb (hunk 3 above)
Comment (zreichert) [inline spec/acceptance/instance_spec.rb]: No problem, I'll condense it.

File: spec/acceptance/instance_spec.rb [hunk 4]
```diff
@@ -231,4 +232,150 @@ def expect_failed_apply(config)
...
+      # re-assign @instance for more up to date info
+      @instance = get_instance(@config[:name])
+    end
+
+    after(:all) do
+      @config[:ensure] = 'absent'
+      TestExecutor.puppet_resource('ec2_instance', @config, '--modulepath ../')
+      @aws.ec2_client.wait_until(:instance_terminated, instance_ids:[@instance.instance_id]) do |w|
+        w.max_attempts = 5
+        # this will fail until CLOUD-202 is resolved
```
Comment (Iristyle) [inline spec/acceptance/instance_spec.rb]: comment can be removed

File: spec/acceptance/instance_spec.rb [hunk 5]
```diff
@@ -231,4 +232,150 @@ def expect_failed_apply(config)
...
+      expect(@result.stdout).to match(regex)
+    end
+
+    it 'image_id is correct' do
+      regex = /(image_id)(\s*)(=>)(\s*)('#{@config[:image_id]}')/
+      expect(@result.stdout).to match(regex)
+    end
+
+    it 'security_groups is correct' do
+      pending('this test is blocked by CLOUD-196')
```
Comment (Iristyle) [inline spec/acceptance/instance_spec.rb]: Which PR is this?  Let's get it merged first...

Comment (zreichert) [inline spec/acceptance/instance_spec.rb]: I don't think that there is a PR for this yet.

Comment (garethr) [inline spec/acceptance/instance_spec.rb]: Its now in https://github.com/puppetlabs/puppetlabs-aws/pull/71

Comment (Iristyle) [inline spec/acceptance/instance_spec.rb]: @zreichert you can unblock this test and update this PR now that #71 is merged.

Thanks!

File: spec/acceptance/instance_spec.rb [hunk 6]
```diff
@@ -231,4 +232,148 @@ def expect_failed_apply(config)
...
+      @config[:ensure] = 'absent'
+      TestExecutor.puppet_resource('ec2_instance', @config, '--modulepath ../')
+      @aws.ec2_client.wait_until(:instance_terminated, instance_ids:[@instance.instance_id]) do |w|
+        w.max_attempts = 5
+      end
+    end
+
+    it 'ensure is correct' do
+      regex = /(ensure)(\s*)(=>)(\s*)('present')/
+      expect(@result.stdout).to match(regex)
```
Comment (zreichert) [inline spec/acceptance/instance_spec.rb]: @Iristyle, @cmurphy, @garethr  should I update this to match 'running' instead of 'present'?

Comment (cmurphy) [inline spec/acceptance/instance_spec.rb]: I think we decided 'running' is more user-friendly and lines up with how the service resource behaves.

Comment (zreichert) [inline spec/acceptance/instance_spec.rb]: I agree, I will update the test.

**Pull request merged**